packages/core/src/new-api/internal/execution/transactions/transaction-service.ts

```diff
@@ -0,0 +1,69 @@
+import { IgnitionError } from "../../../../errors";
+import { ArtifactResolver } from "../../../types/artifact";
+import {
+  DeployContractInteraction,
+  DeployContractResult,
+  OnchainInteraction,
+  OnchainResult,
+} from "../../../types/journal";
+import { TransactionService } from "../../../types/transaction-service";
+import { isDeployContractInteraction } from "../guards";
+
+import { ChainDispatcher } from "./chain-dispatcher";
+
+/**
+ * A service for managing on-chain interactions during the deployment.
+ *
+ * @beta
+ */
+export class TransactionServiceImplementation implements TransactionService {
+  constructor(
+    private _artifactLoader: ArtifactResolver,
+    private _chainDispatcher: ChainDispatcher
+  ) {}
+
+  public async onchain(
+    interaction: OnchainInteraction
+  ): Promise<OnchainResult> {
+    if (!isDeployContractInteraction(interaction)) {
+      throw new IgnitionError(
+        "Transaction service not implemented for this interaction"
+      );
+    }
+
+    return this._dispatchDeployContract(interaction);
+  }
+
+  private async _dispatchDeployContract(
+    deployContractInteraction: DeployContractInteraction
+  ): Promise<DeployContractResult> {
+    // TODO: consider replacing this with a registry of artifacts
+    const artifact = await this._artifactLoader.load(
```

Comment:
I think this PR misses the fact that we need two artifact adapters. One for the project's, one for the deployment's artifacts.

Reply:
Also, we can't fit a registry of artifacts in memory, as mentioned in the design doc.

Reply:
After our discussions, a deployment loader that contains and exposes the journal will be added as part of https://github.com/NomicFoundation/ignition/issues/109